languages/python/sqlalchemy-oso/sqlalchemy_oso/enforcer.py

```diff
@@ -0,0 +1,50 @@
+from oso import Enforcer
+from .oso import SQLAlchemyPolicy
+from .auth import authorize_model
+
+
+class SQLAlchemyEnforcer(Enforcer):
+    """
+    NOTE: This is a preview feature.
+
+    Custom Oso enforcer for SQLAlchemy.
+    """
+
+    def __init__(self, policy: SQLAlchemyPolicy, session_maker, *args, **kwargs):
+        """Construct a new SQLAlchemyEnforcer
+
+        > policy = SQLAlchemyPolicy(Base)
+        > oso = SQLAlchemyEnforcer(policy, Session)
+        > # ...
+        > oso.authorize_query(user, Article).all()
+
+        :param policy: An instance of ``SQLAlchemyPolicy``
+        :param session_maker: A SQLAlchemy session maker instance"""
+        self.session_maker = session_maker
+        super().__init__(policy, *args, **kwargs)
+
+    def authorize_query(self, actor, model, action=None):
```

Comment:
We intentionally don't expose a query authorization API for SQLAlchemy currently. `authorize_model` is internal and only used on the session API.

Authorization is associated with the session because SQLAlchemy may issue queries under the hood, and because sessions cache objects that have already been retrieved from the database.

For example, consider this policy:

```
allow(user: User, "read", post: Post) if
   post.owner = user;

allow(user: User, "read", _post_analytics: PostAnalytics) if
    user.is_admin;
```

Assume the current user is not an admin:

```
enforcer = SQLAlchemyEnforcer(session=session)
posts = enforcer.authorize_query(user, Post).all()

analytics = [p.analytics for p in posts]
return json.dumps(analytics)
```

This would return analytics data to a user, because accessing the `p.analytics` property issues a query to the database under the hood. This can unintentionally expose data that should be authorized.

With the current API, authorization is attached to a session.

```
session = AuthorizedSession(oso, user, {Post: "read", PostAnalytics: "read"})
posts = session.query(Post).all()

analytics = [p.analytics for p in posts]
# analytics will be an array of [None, None, None, ... ] since the user isn't authorized to read this data
```

I don't think the current API is ideal either, we've gotten some feedback that it's non-intuitive. But, it does prevent unintentionally fetching objects without applying authorization.

Reply:
Mostly resolved this offline, but I am curious whether you think it makes more sense to directly return the result of `authorize_model` instead of attaching it to a session. That's a bit of a sqlalchemy style question, so I might defer to you.

The current API I have is:
```
query = oso.authorize_query(user, Expense)
query.all()
```

We might also consider something like this (which might be more sqlalchemy-y):
```
query = oso.authorize_query(user, Expense)
db.session.query(Expense).filter(query).all()
```

Do you have a preference?

Reply:
Not really. Although one advantage of the latter is we wouldn't need the Enforcer to know how to construct a session.

We could also achieve that by passing the session in to `authorize_query`, or passing the query object itself into `authorize_query` (which would already have the session associated with it).

Reply:
I tried out this other API last night, but given the way that `authorize_model` works, we need the session anyway. I'm going to merge this as-is, but we should definitely revisit this once we're using the new data filtering methods.